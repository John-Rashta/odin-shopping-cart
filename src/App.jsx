import { useState, useEffect } from 'react';
import { Header } from './components/Header';
import { HomePage } from './components/HomePage';
import { ShoppingPage } from './components/ShoppingPage';
import { useParams } from 'react-router-dom';
import styled from 'styled-components';



function App() {
  const [shopData, setShopData] = useState([]);
  const [shopCart, setShopCart] = useState([]);

  useEffect(() => {
      const fetchDataForShop = async () => {
          try {
              const rawResponse = await fetch('https://fakestoreapi.com/products');
              
              if (!rawResponse.ok) {
                  throw new Error(`HTTP error: Status ${rawResponse.status}`);
              }

              let fullData = await rawResponse.json();
              const newData = [];
              fullData.forEach((val) => {
                  newData.push({...val});
              });

              setShopData(fullData);

          } catch (err) {
              console.log(err.message);
          }
      }

      fetchDataForShop();
  }, []);

  function addToCart(amount, id, price, name) {
      if (shopCart.some((val) => val[0] === id)) {
          const rawCart = shopCart;
          const newCart = rawCart.map((val) => {
              if (val[0] === id) {
                  return [val[0], Number(val[1]) + amount, val[2], val[3]];
              }
              return val;
          });
          setShopCart(newCart);
          return;
      }
      const newCart = shopCart;
      newCart.push([id, Number(amount), price, name]);
      setShopCart(newCart);
  }

  function deleteCart(id) {
    const rawCart = shopCart;
    const newCart = rawCart.filter((val) => val[0] !== id);
    setShopCart(newCart);
  }

  function changeCart(id, amount) {
    const rawCart = shopCart;
    const newCart = rawCart.reduce((prevArr, currentItem) => {
      if (currentItem[0] === id) {
        if (currentItem[1] + amount <= 0) {
          return prevArr;
        }
          prevArr.push([currentItem[0], Number(currentItem[1]) + amount, currentItem[2], currentItem[3]]);
          return prevArr;
      }
      prevArr.push(currentItem);
      return prevArr;
    }, []);
    setShopCart(newCart);
  }

  const {name} = useParams();

  return (
    <MainContainer>
        <Header data={{shopCart, shopData}} deleteCart={deleteCart} changeCart={changeCart} />
        {
          name === "shop" ? (
            <ShoppingPage fullData={{shopData, shopCart}} addToCart={addToCart} />
          ) : (
            <HomePage/>
          )
        }

    </MainContainer>
    
  )
}

const MainContainer = styled.div`
  height: 100%;
  width: 100%;
  min-height: 100vh;
  background-color: #fbcfe8;
`;

export default App;
