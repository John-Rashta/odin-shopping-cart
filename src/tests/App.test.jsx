import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import App from '../App.jsx';
import {BrowserRouter} from 'react-router-dom';
import userEvent from "@testing-library/user-event";

describe("App component", () => {
    it("renders Homepage properly", () => {
      const container = render(<App />, {wrapper: BrowserRouter});
      expect(container).toMatchSnapshot();
    }); 

    it("renders cart properly", async () => {
        render(<App />, {wrapper: BrowserRouter});
        const user = userEvent.setup();
        const showButton = screen.getByText("Cart");

        await user.click(showButton);

        expect(screen.getByText("No Items").textContent).toMatch("No Items");

    }); 
});