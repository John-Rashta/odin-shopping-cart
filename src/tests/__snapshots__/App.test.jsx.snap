// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders Homepage properly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jhZTHU euVtHf"
      >
        <header
          class="sc-dstKZu dHBwHX"
        >
          <div
            class="sc-keTIit eSzApg"
          >
            <a
              data-discover="true"
              href="/"
              style="color: rgb(29, 78, 216);"
            >
              Home
            </a>
            <a
              data-discover="true"
              href="/shop"
              style="color: rgb(29, 78, 216);"
            >
              Shop
            </a>
          </div>
          <div
            class="sc-jwIPbr fyKCMZ"
          >
            Something Shop
          </div>
          <div
            class="sc-dpBQxM kiFWGY"
          >
            <div
              class="sc-keTIit eSzApg showCartButton"
            >
              <svg
                class="showCartButton"
                role="presentation"
                style="width: 1.5rem; height: 1.5rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17,18C15.89,18 15,18.89 15,20A2,2 0 0,0 17,22A2,2 0 0,0 19,20C19,18.89 18.1,18 17,18M1,2V4H3L6.6,11.59L5.24,14.04C5.09,14.32 5,14.65 5,15A2,2 0 0,0 7,17H19V15H7.42A0.25,0.25 0 0,1 7.17,14.75C7.17,14.7 7.18,14.66 7.2,14.63L8.1,13H15.55C16.3,13 16.96,12.58 17.3,11.97L20.88,5.5C20.95,5.34 21,5.17 21,5A1,1 0 0,0 20,4H5.21L4.27,2M7,18C5.89,18 5,18.89 5,20A2,2 0 0,0 7,22A2,2 0 0,0 9,20C9,18.89 8.1,18 7,18Z"
                  style="fill: currentColor;"
                />
              </svg>
              <div>
                Cart
              </div>
            </div>
          </div>
        </header>
        <main
          class="sc-cHqXqK fKnaqh"
        >
          <h1
            class="sc-ovuCP ecMAHf"
          >
            Welcome to Something Something Shop
          </h1>
          <h2
            class="sc-jtQUzJ gSOomK"
          >
            Enjoy your Shopping!
          </h2>
          <div
            class="sc-cEzcPc fJcLT"
          >
            <a
              data-discover="true"
              href="/shop"
              style="font-size: 1.5rem; color: rgb(109, 40, 217);"
            >
              Start Shopping
            </a>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jhZTHU euVtHf"
    >
      <header
        class="sc-dstKZu dHBwHX"
      >
        <div
          class="sc-keTIit eSzApg"
        >
          <a
            data-discover="true"
            href="/"
            style="color: rgb(29, 78, 216);"
          >
            Home
          </a>
          <a
            data-discover="true"
            href="/shop"
            style="color: rgb(29, 78, 216);"
          >
            Shop
          </a>
        </div>
        <div
          class="sc-jwIPbr fyKCMZ"
        >
          Something Shop
        </div>
        <div
          class="sc-dpBQxM kiFWGY"
        >
          <div
            class="sc-keTIit eSzApg showCartButton"
          >
            <svg
              class="showCartButton"
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M17,18C15.89,18 15,18.89 15,20A2,2 0 0,0 17,22A2,2 0 0,0 19,20C19,18.89 18.1,18 17,18M1,2V4H3L6.6,11.59L5.24,14.04C5.09,14.32 5,14.65 5,15A2,2 0 0,0 7,17H19V15H7.42A0.25,0.25 0 0,1 7.17,14.75C7.17,14.7 7.18,14.66 7.2,14.63L8.1,13H15.55C16.3,13 16.96,12.58 17.3,11.97L20.88,5.5C20.95,5.34 21,5.17 21,5A1,1 0 0,0 20,4H5.21L4.27,2M7,18C5.89,18 5,18.89 5,20A2,2 0 0,0 7,22A2,2 0 0,0 9,20C9,18.89 8.1,18 7,18Z"
                style="fill: currentColor;"
              />
            </svg>
            <div>
              Cart
            </div>
          </div>
        </div>
      </header>
      <main
        class="sc-cHqXqK fKnaqh"
      >
        <h1
          class="sc-ovuCP ecMAHf"
        >
          Welcome to Something Something Shop
        </h1>
        <h2
          class="sc-jtQUzJ gSOomK"
        >
          Enjoy your Shopping!
        </h2>
        <div
          class="sc-cEzcPc fJcLT"
        >
          <a
            data-discover="true"
            href="/shop"
            style="font-size: 1.5rem; color: rgb(109, 40, 217);"
          >
            Start Shopping
          </a>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
